contracts/BSCValidatorSet.sol

```diff
@@ -470,15 +557,182 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
+      for (uint i = index + 1; i < n; i++) {
+        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
+      }
+    }
+    // averageDistribute*rest may less than income, but it is ok, the dust income will go to system reward eventually.
+
+    return index;
+  }
+
+  function _felony(address validator) private {
+    uint256 index = currentValidatorSetMap[validator];
+    if (index <= 0) {
+      return;
+    }
+    // the actually index
+    index = index - 1;
+    uint256 income = currentValidatorSet[index].incoming;
+    uint256 rest = currentValidatorSet.length - 1;
+    if (getValidators().length <= 1) {
+      // will not remove the validator if it is the only one validator.
+      currentValidatorSet[index].incoming = 0;
+      return;
+    }
+    emit validatorFelony(validator, income);
+
+    // remove the validator from currentValidatorSet
+    delete currentValidatorSetMap[validator];
+    // It is ok that the validatorSet is not in order.
+    if (index != currentValidatorSet.length - 1) {
+      currentValidatorSet[index] = currentValidatorSet[currentValidatorSet.length - 1];
+      validatorExtraSet[index] = validatorExtraSet[currentValidatorSet.length - 1];
+
+      currentValidatorSetMap[currentValidatorSet[index].consensusAddress] = index + 1;
+    }
+    currentValidatorSet.pop();
+    validatorExtraSet.pop();
+
+    uint256 averageDistribute = income / rest;
+    if (averageDistribute != 0) {
+      uint n = currentValidatorSet.length;
+      for (uint i = 0; i < n; i++) {
+        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
+      }
+    }
+    // averageDistribute*rest may less than income, but it is ok, the dust income will go to system reward eventually.
+  }
+
+  function _forceMaintainingValidatorsExit(Validator[] memory _validatorSet) private returns (Validator[] memory unjailedValidatorSet){
+    uint256 numOfFelony = 0;
+    address validator;
+    bool isFelony;
+
+    // 1. validators exit maintenance
+    for (uint i = 0; i < currentValidatorSet.length; i++) {
+      if (!validatorExtraSet[i].isMaintaining) {
+        continue;
+      }
+
+      // only maintaining validators
+      validator = currentValidatorSet[i].consensusAddress;
+
+      // exit maintenance
+      isFelony = _exitMaintenance(validator, i);
+      if (!isFelony || numOfFelony >= _validatorSet.length - 1) {
+        continue;
+      }
+
+      // record the jailed validator in validatorSet
+      for (uint index = 0; index < _validatorSet.length; index++) {
+        if (_validatorSet[index].consensusAddress == validator) {
+          _validatorSet[index].jailed = true;
+          numOfFelony++;
+          break;
+        }
+      }
+    }
+
+    // 2. clear all maintain info
+    for (uint i = 0; i < currentValidatorSet.length; i++) {
+      validatorExtraSet[i].isMaintaining = false;
+      validatorExtraSet[i].enterMaintenanceHeight = 0;
```

Comment:
Duplicated, we can move all this into _exitMaintenance().

Reply:
`validatorExtraSet[i].enterMaintenanceHeight = 0;` should not be moved into  `_exitMaintenance().`.
Consider this, how do we check whether a validator entered and exited voluntarily `Temporary Maintenance`  yet in single 24-hours period ?  - Check whether `validatorExtraSet[i].enterMaintenanceHeight > 0`  in this condition

Reply:
Make sense